pkg/watch/health.go

```diff
@@ -0,0 +1,239 @@
+	reality, _, err := store.ListPods(path)
+	if err != nil {
+		logger.WithField("inner_err", err).Warningln("failed to get pods from reality store")
+	}
+
+	return updatePods(watchedPods, reality, logger, store, node)
+}
+
+func resultFromCheck(resp *http.Response) (health.Result, error) {
+	var err error
+	res := health.Result{}
+	if resp == nil {
+		res.Status = health.Critical
+		return res, nil
+	}
+	if resp.StatusCode == 200 {
+		res.Status = health.Passing
+	} else if resp.StatusCode > 200 && resp.StatusCode < 300 {
+		res.Status = health.Passing
+		res.Output, err = getBody(resp)
+	} else {
+		res.Status = health.Critical
+		res.Output, err = getBody(resp)
+	}
+	return res, err
+}
+
+func getBody(resp *http.Response) (string, error) {
+	defer resp.Body.Close()
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return "", err
+	}
+	return string(body), nil
+}
+
+// once we get health data we need to make a put request
+// to consul to put the data in the KV Store
+func writeToConsul(res health.Result, store kp.Store) (time.Time, error) {
+	timeOfPut, _, err := store.PutHealth(resToKPRes(res))
+	return timeOfPut, err
+}
+
+func resToKPRes(res health.Result) kp.WatchResult {
+	return kp.WatchResult{
+		Service: res.Service,
+		Node:    res.Node,
+		Id:      res.ID,
+		Status:  string(res.Status),
+		Output:  res.Output,
+	}
+}
+
+func (p *PodWatch) updateNeeded(res health.Result, ttl time.Duration) bool {
+	// if status has changed indicate that consul needs to be updated
+	if p.lastStatus != res.Status {
+		return true
+	}
+	// if more than TTL / 4 seconds have elapsed since previous check
+	// indicate that consul needs to be updated
+	if time.Since(p.lastCheck) > time.Duration(ttl/4)*time.Second {
+		return true
+	}
+
+	return false
+}
+
+// compares services being monitored with services that
+// need to be monitored.
+func updatePods(current []PodWatch, reality []kp.ManifestResult, logger *logging.Logger, store kp.Store, node string) []PodWatch {
+	newCurrent := []PodWatch{}
+	// for pod in current if pod not in reality: kill
+	for _, pod := range current {
+		inReality := false
+		for _, man := range reality {
+			if man.Manifest.Id == pod.manifest.Id {
+				inReality = true
+				break
+			}
+		}
+
+		// if this podwatch is not in the reality store kill its go routine
+		// else add this podwatch to newCurrent
+		if inReality == false {
+			pod.shutdownCh <- true
+		} else {
+			newCurrent = append(newCurrent, pod)
+		}
+	}
+	// for pod in reality if pod not in current: create podwatch and
+	// append to current
+	for _, man := range reality {
+		missing := true
+		for _, pod := range newCurrent {
+			if man.Manifest.Id == pod.manifest.Id {
+				missing = false
+				break
+			}
+		}
+
+		// if a manifest is in reality but not current a podwatch is created
+		// with that manifest and added to newCurrent
+		if missing == true {
+			newPod := PodWatch{
+				manifest:   man.Manifest,
+				shutdownCh: make(chan bool, 1),
```

Comment:
I think my previous incarnation of this question got erased: why use a buffered channel here?